Failed to Download Go language server

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Opening a go project starts to download gopls, but this fails after some retries.

Using Go v1.21

Note: Even though the logs state that the language server has started before failing to download, this is not true, and I do not have LSP features.


### Environment

Zed: v0.121.5 (Zed)
OS: macOS 14.3.0
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

```
2024-02-08T00:14:23+00:00 [INFO] open paths ["/Users/andre/.config/zed/settings.json"]
2024-02-08T00:14:23+00:00 [INFO] Initializing default prettier with plugins {}
2024-02-08T00:14:23+00:00 [INFO] starting language server "json-language-server", path: "/Users/andre/.config/zed", id: 4
2024-02-08T00:14:23+00:00 [INFO] Node runtime install_if_needed
2024-02-08T00:14:24+00:00 [INFO] Node runtime install_if_needed
2024-02-08T00:14:25+00:00 [INFO] 0 unhandled notification LogMessage:
{
  "level": 1,
  "message": "[INFO] [default] [2024-02-08T00:14:25.999Z] [fetchCompletions] engine https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex",
  "metadataStr": "[INFO] [default] [2024-02-08T00:14:25.999Z]",
  "extra": [
    "[fetchCompletions] engine https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex"
  ]
}
2024-02-08T00:14:26+00:00 [INFO] 0 unhandled notification LogMessage:
{
  "level": 1,
  "message": "[INFO] [default] [2024-02-08T00:14:26.101Z] request.response: [https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex/completions] took 102 ms",
  "metadataStr": "[INFO] [default] [2024-02-08T00:14:26.101Z]",
  "extra": [
2024-02-08T00:56:27+00:00 [ERROR] server stderr: Some("")
2024-02-08T00:56:27+00:00 [ERROR] crates/zed/src/languages/go.rs:357: no cached binary
2024-02-08T00:56:27+00:00 [INFO] About to spawn test binary
2024-02-08T00:56:27+00:00 [WARN] test binary failed to launch
2024-02-08T00:56:27+00:00 [WARN] test binary check failed
2024-02-08T00:56:27+00:00 [INFO] beginning to reinstall server
2024-02-08T00:56:27+00:00 [INFO] deleting server container
2024-02-08T00:56:27+00:00 [INFO] starting language server "gopls", path: "/Users/andre/github/Quasarch/quasarplane", id: 3
2024-02-08T00:56:27+00:00 [ERROR] crates/zed/src/languages/go.rs:357: no cached binary
2024-02-08T00:56:27+00:00 [ERROR] failed to start language server "gopls": Failed to find a release
2024-02-08T00:56:27+00:00 [ERROR] server stderr: Some("")
2024-02-08T00:56:27+00:00 [ERROR] crates/zed/src/languages/go.rs:357: no cached binary
2024-02-08T00:56:27+00:00 [INFO] About to spawn test binary
2024-02-08T00:56:27+00:00 [WARN] test binary failed to launch
2024-02-08T00:56:27+00:00 [WARN] test binary check failed
2024-02-08T00:56:27+00:00 [INFO] beginning to reinstall server
2024-02-08T00:56:27+00:00 [INFO] deleting server container
2024-02-08T00:56:27+00:00 [INFO] starting language server "gopls", path: "/Users/andre/github/Quasarch/quasarplane", id: 4
2024-02-08T00:56:27+00:00 [ERROR] crates/zed/src/languages/go.rs:357: no cached binary
2024-02-08T00:56:27+00:00 [ERROR] failed to start language server "gopls": Failed to find a release
2024-02-08T00:56:27+00:00 [ERROR] server stderr: Some("")
2024-02-08T00:56:27+00:00 [ERROR] crates/zed/src/languages/go.rs:357: no cached binary
2024-02-08T00:56:27+00:00 [INFO] About to spawn test binary
2024-02-08T00:56:27+00:00 [WARN] test binary failed to launch
2024-02-08T00:56:27+00:00 [WARN] test binary check failed
2024-02-08T00:56:27+00:00 [INFO] beginning to reinstall server
2024-02-08T00:56:27+00:00 [INFO] deleting server container
2024-02-08T00:56:27+00:00 [INFO] starting language server "gopls", path: "/Users/andre/github/Quasarch/quasarplane", id: 5
2024-02-08T00:56:27+00:00 [INFO] 0 unhandled notification LogMessage:
{
  "level": 0,
  "message": "[DEBUG] [agent] [2024-02-08T00:56:27.715Z] Telemetry initialized",
  "metadataStr": "[DEBUG] [agent] [2024-02-08T00:56:27.715Z]",
  "extra": [
    "Telemetry initialized"
  ]
}
2024-02-08T00:56:27+00:00 [ERROR] crates/zed/src/languages/go.rs:357: no cached binary
2024-02-08T00:56:27+00:00 [ERROR] failed to start language server "gopls": Failed to find a release
2024-02-08T00:56:27+00:00 [ERROR] server stderr: Some("")
2024-02-08T00:56:27+00:00 [ERROR] Hit 4 reinstallation attempts for "gopls"
```

--- a/crates/copilot/src/copilot.rs
+++ b/crates/copilot/src/copilot.rs
@@ -976,7 +976,8 @@ async fn get_copilot_lsp(http: Arc<dyn HttpClient>) -> anyhow::Result<PathBuf> {
 
     ///Check for the latest copilot language server and download it if we haven't already
     async fn fetch_latest(http: Arc<dyn HttpClient>) -> anyhow::Result<PathBuf> {
-        let release = latest_github_release("zed-industries/copilot", false, http.clone()).await?;
+        let release =
+            latest_github_release("zed-industries/copilot", false, false, http.clone()).await?;
 
         let version_dir = &*paths::COPILOT_DIR.join(format!("copilot-{}", release.name));
 

--- a/crates/copilot/src/copilot.rs
+++ b/crates/copilot/src/copilot.rs
@@ -977,7 +977,7 @@ async fn get_copilot_lsp(http: Arc<dyn HttpClient>) -> anyhow::Result<PathBuf> {
     ///Check for the latest copilot language server and download it if we haven't already
     async fn fetch_latest(http: Arc<dyn HttpClient>) -> anyhow::Result<PathBuf> {
         let release =
-            latest_github_release("zed-industries/copilot", false, false, http.clone()).await?;
+            latest_github_release("zed-industries/copilot", true, false, http.clone()).await?;
 
         let version_dir = &*paths::COPILOT_DIR.join(format!("copilot-{}", release.name));
 
@@ -998,7 +998,7 @@ async fn get_copilot_lsp(http: Arc<dyn HttpClient>) -> anyhow::Result<PathBuf> {
             let mut response = http
                 .get(url, Default::default(), true)
                 .await
-                .map_err(|err| anyhow!("error downloading copilot release: {}", err))?;
+                .context("error downloading copilot release")?;
             let decompressed_bytes = GzipDecoder::new(BufReader::new(response.body_mut()));
             let archive = Archive::new(decompressed_bytes);
             archive.unpack(dist_dir).await?;

--- a/crates/util/src/github.rs
+++ b/crates/util/src/github.rs
@@ -27,6 +27,7 @@ pub struct GithubReleaseAsset {
 
 pub async fn latest_github_release(
     repo_name_with_owner: &str,
+    require_assets: bool,
     pre_release: bool,
     http: Arc<dyn HttpClient>,
 ) -> Result<GithubRelease, anyhow::Error> {
@@ -68,6 +69,7 @@ pub async fn latest_github_release(
 
     releases
         .into_iter()
-        .find(|release| !release.assets.is_empty() && release.pre_release == pre_release)
+        .filter(|release| !require_assets || !release.assets.is_empty())
+        .find(|release| release.pre_release == pre_release)
         .ok_or(anyhow!("Failed to find a release"))
 }

--- a/crates/zed/src/languages/c.rs
+++ b/crates/zed/src/languages/c.rs
@@ -29,7 +29,7 @@ impl super::LspAdapter for CLspAdapter {
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
         let release =
-            latest_github_release("clangd/clangd", false, false, delegate.http_client()).await?;
+            latest_github_release("clangd/clangd", true, false, delegate.http_client()).await?;
         let asset_name = format!("clangd-mac-{}.zip", release.name);
         let asset = release
             .assets

--- a/crates/zed/src/languages/c.rs
+++ b/crates/zed/src/languages/c.rs
@@ -28,7 +28,8 @@ impl super::LspAdapter for CLspAdapter {
         &self,
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
-        let release = latest_github_release("clangd/clangd", false, delegate.http_client()).await?;
+        let release =
+            latest_github_release("clangd/clangd", false, false, delegate.http_client()).await?;
         let asset_name = format!("clangd-mac-{}.zip", release.name);
         let asset = release
             .assets

--- a/crates/zed/src/languages/csharp.rs
+++ b/crates/zed/src/languages/csharp.rs
@@ -29,14 +29,6 @@ impl super::LspAdapter for OmniSharpAdapter {
         &self,
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
-        let release = latest_github_release(
-            "OmniSharp/omnisharp-roslyn",
-            false,
-            false,
-            delegate.http_client(),
-        )
-        .await?;
-
         let mapped_arch = match ARCH {
             "aarch64" => Some("arm64"),
             "x86_64" => Some("x64"),
@@ -46,6 +38,13 @@ impl super::LspAdapter for OmniSharpAdapter {
         match mapped_arch {
             None => Ok(Box::new(())),
             Some(arch) => {
+                let release = latest_github_release(
+                    "OmniSharp/omnisharp-roslyn",
+                    true,
+                    false,
+                    delegate.http_client(),
+                )
+                .await?;
                 let asset_name = format!("omnisharp-osx-{}-net6.0.tar.gz", arch);
                 let asset = release
                     .assets

--- a/crates/zed/src/languages/csharp.rs
+++ b/crates/zed/src/languages/csharp.rs
@@ -29,9 +29,13 @@ impl super::LspAdapter for OmniSharpAdapter {
         &self,
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
-        let release =
-            latest_github_release("OmniSharp/omnisharp-roslyn", false, delegate.http_client())
-                .await?;
+        let release = latest_github_release(
+            "OmniSharp/omnisharp-roslyn",
+            false,
+            false,
+            delegate.http_client(),
+        )
+        .await?;
 
         let mapped_arch = match ARCH {
             "aarch64" => Some("arm64"),

--- a/crates/zed/src/languages/deno.rs
+++ b/crates/zed/src/languages/deno.rs
@@ -70,7 +70,8 @@ impl LspAdapter for DenoLspAdapter {
         &self,
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
-        let release = latest_github_release("denoland/deno", false, delegate.http_client()).await?;
+        let release =
+            latest_github_release("denoland/deno", false, false, delegate.http_client()).await?;
         let asset_name = format!("deno-{}-apple-darwin.zip", consts::ARCH);
         let asset = release
             .assets

--- a/crates/zed/src/languages/deno.rs
+++ b/crates/zed/src/languages/deno.rs
@@ -71,7 +71,7 @@ impl LspAdapter for DenoLspAdapter {
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
         let release =
-            latest_github_release("denoland/deno", false, false, delegate.http_client()).await?;
+            latest_github_release("denoland/deno", true, false, delegate.http_client()).await?;
         let asset_name = format!("deno-{}-apple-darwin.zip", consts::ARCH);
         let asset = release
             .assets

--- a/crates/zed/src/languages/elixir.rs
+++ b/crates/zed/src/languages/elixir.rs
@@ -111,7 +111,7 @@ impl LspAdapter for ElixirLspAdapter {
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
         let http = delegate.http_client();
-        let release = latest_github_release("elixir-lsp/elixir-ls", false, http).await?;
+        let release = latest_github_release("elixir-lsp/elixir-ls", false, false, http).await?;
         let version_name = release
             .name
             .strip_prefix("Release ")
@@ -314,7 +314,8 @@ impl LspAdapter for NextLspAdapter {
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
         let release =
-            latest_github_release("elixir-tools/next-ls", false, delegate.http_client()).await?;
+            latest_github_release("elixir-tools/next-ls", false, false, delegate.http_client())
+                .await?;
         let version = release.name.clone();
         let platform = match consts::ARCH {
             "x86_64" => "darwin_amd64",

--- a/crates/zed/src/languages/elixir.rs
+++ b/crates/zed/src/languages/elixir.rs
@@ -111,19 +111,19 @@ impl LspAdapter for ElixirLspAdapter {
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
         let http = delegate.http_client();
-        let release = latest_github_release("elixir-lsp/elixir-ls", false, false, http).await?;
+        let release = latest_github_release("elixir-lsp/elixir-ls", true, false, http).await?;
         let version_name = release
             .name
             .strip_prefix("Release ")
             .context("Elixir-ls release name does not start with prefix")?
             .to_owned();
 
-        let asset_name = format!("elixir-ls-{}.zip", &version_name);
+        let asset_name = format!("elixir-ls-{version_name}.zip");
         let asset = release
             .assets
             .iter()
             .find(|asset| asset.name == asset_name)
-            .ok_or_else(|| anyhow!("no asset found matching {:?}", asset_name))?;
+            .ok_or_else(|| anyhow!("no asset found matching {asset_name:?}"))?;
 
         let version = GitHubLspBinaryVersion {
             name: version_name,
@@ -313,21 +313,21 @@ impl LspAdapter for NextLspAdapter {
         &self,
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
-        let release =
-            latest_github_release("elixir-tools/next-ls", false, false, delegate.http_client())
-                .await?;
-        let version = release.name.clone();
         let platform = match consts::ARCH {
             "x86_64" => "darwin_amd64",
             "aarch64" => "darwin_arm64",
             other => bail!("Running on unsupported platform: {other}"),
         };
-        let asset_name = format!("next_ls_{}", platform);
+        let release =
+            latest_github_release("elixir-tools/next-ls", true, false, delegate.http_client())
+                .await?;
+        let version = release.name;
+        let asset_name = format!("next_ls_{platform}");
         let asset = release
             .assets
             .iter()
             .find(|asset| asset.name == asset_name)
-            .ok_or_else(|| anyhow!("no asset found matching {:?}", asset_name))?;
+            .with_context(|| format!("no asset found matching {asset_name:?}"))?;
         let version = GitHubLspBinaryVersion {
             name: version,
             url: asset.browser_download_url.clone(),

--- a/crates/zed/src/languages/gleam.rs
+++ b/crates/zed/src/languages/gleam.rs
@@ -35,7 +35,7 @@ impl LspAdapter for GleamLspAdapter {
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
         let release =
-            latest_github_release("gleam-lang/gleam", false, false, delegate.http_client()).await?;
+            latest_github_release("gleam-lang/gleam", true, false, delegate.http_client()).await?;
 
         let asset_name = format!(
             "gleam-{version}-{arch}-apple-darwin.tar.gz",

--- a/crates/zed/src/languages/gleam.rs
+++ b/crates/zed/src/languages/gleam.rs
@@ -35,7 +35,7 @@ impl LspAdapter for GleamLspAdapter {
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
         let release =
-            latest_github_release("gleam-lang/gleam", false, delegate.http_client()).await?;
+            latest_github_release("gleam-lang/gleam", false, false, delegate.http_client()).await?;
 
         let asset_name = format!(
             "gleam-{version}-{arch}-apple-darwin.tar.gz",

--- a/crates/zed/src/languages/go.rs
+++ b/crates/zed/src/languages/go.rs
@@ -45,7 +45,8 @@ impl super::LspAdapter for GoLspAdapter {
         &self,
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
-        let release = latest_github_release("golang/tools", false, delegate.http_client()).await?;
+        let release =
+            latest_github_release("golang/tools", false, false, delegate.http_client()).await?;
         let version: Option<String> = release.name.strip_prefix("gopls/v").map(str::to_string);
         if version.is_none() {
             log::warn!(

--- a/crates/zed/src/languages/lua.rs
+++ b/crates/zed/src/languages/lua.rs
@@ -30,19 +30,19 @@ impl super::LspAdapter for LuaLspAdapter {
         &self,
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
-        let release = latest_github_release(
-            "LuaLS/lua-language-server",
-            false,
-            false,
-            delegate.http_client(),
-        )
-        .await?;
-        let version = release.name.clone();
         let platform = match consts::ARCH {
             "x86_64" => "x64",
             "aarch64" => "arm64",
             other => bail!("Running on unsupported platform: {other}"),
         };
+        let release = latest_github_release(
+            "LuaLS/lua-language-server",
+            true,
+            false,
+            delegate.http_client(),
+        )
+        .await?;
+        let version = &release.name;
         let asset_name = format!("lua-language-server-{version}-darwin-{platform}.tar.gz");
         let asset = release
             .assets
@@ -50,7 +50,7 @@ impl super::LspAdapter for LuaLspAdapter {
             .find(|asset| asset.name == asset_name)
             .ok_or_else(|| anyhow!("no asset found matching {:?}", asset_name))?;
         let version = GitHubLspBinaryVersion {
-            name: release.name.clone(),
+            name: release.name,
             url: asset.browser_download_url.clone(),
         };
         Ok(Box::new(version) as Box<_>)

--- a/crates/zed/src/languages/lua.rs
+++ b/crates/zed/src/languages/lua.rs
@@ -30,9 +30,13 @@ impl super::LspAdapter for LuaLspAdapter {
         &self,
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
-        let release =
-            latest_github_release("LuaLS/lua-language-server", false, delegate.http_client())
-                .await?;
+        let release = latest_github_release(
+            "LuaLS/lua-language-server",
+            false,
+            false,
+            delegate.http_client(),
+        )
+        .await?;
         let version = release.name.clone();
         let platform = match consts::ARCH {
             "x86_64" => "x64",

--- a/crates/zed/src/languages/rust.rs
+++ b/crates/zed/src/languages/rust.rs
@@ -31,8 +31,13 @@ impl LspAdapter for RustLspAdapter {
         &self,
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
-        let release =
-            latest_github_release("rust-lang/rust-analyzer", false, delegate.http_client()).await?;
+        let release = latest_github_release(
+            "rust-lang/rust-analyzer",
+            false,
+            false,
+            delegate.http_client(),
+        )
+        .await?;
         let asset_name = format!("rust-analyzer-{}-apple-darwin.gz", consts::ARCH);
         let asset = release
             .assets

--- a/crates/zed/src/languages/rust.rs
+++ b/crates/zed/src/languages/rust.rs
@@ -33,7 +33,7 @@ impl LspAdapter for RustLspAdapter {
     ) -> Result<Box<dyn 'static + Send + Any>> {
         let release = latest_github_release(
             "rust-lang/rust-analyzer",
-            false,
+            true,
             false,
             delegate.http_client(),
         )

--- a/crates/zed/src/languages/toml.rs
+++ b/crates/zed/src/languages/toml.rs
@@ -26,7 +26,8 @@ impl LspAdapter for TaploLspAdapter {
         &self,
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
-        let release = latest_github_release("tamasfe/taplo", false, delegate.http_client()).await?;
+        let release =
+            latest_github_release("tamasfe/taplo", true, false, delegate.http_client()).await?;
         let asset_name = format!("taplo-full-darwin-{arch}.gz", arch = std::env::consts::ARCH);
 
         let asset = release

--- a/crates/zed/src/languages/typescript.rs
+++ b/crates/zed/src/languages/typescript.rs
@@ -246,8 +246,13 @@ impl LspAdapter for EsLintLspAdapter {
         // At the time of writing the latest vscode-eslint release was released in 2020 and requires
         // special custom LSP protocol extensions be handled to fully initialize. Download the latest
         // prerelease instead to sidestep this issue
-        let release =
-            latest_github_release("microsoft/vscode-eslint", true, delegate.http_client()).await?;
+        let release = latest_github_release(
+            "microsoft/vscode-eslint",
+            false,
+            false,
+            delegate.http_client(),
+        )
+        .await?;
         Ok(Box::new(GitHubLspBinaryVersion {
             name: release.name,
             url: release.tarball_url,

--- a/crates/zed/src/languages/zig.rs
+++ b/crates/zed/src/languages/zig.rs
@@ -28,7 +28,8 @@ impl LspAdapter for ZlsAdapter {
         &self,
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
-        let release = latest_github_release("zigtools/zls", false, delegate.http_client()).await?;
+        let release =
+            latest_github_release("zigtools/zls", false, false, delegate.http_client()).await?;
         let asset_name = format!("zls-{}-macos.tar.gz", ARCH);
         let asset = release
             .assets

--- a/crates/zed/src/languages/zig.rs
+++ b/crates/zed/src/languages/zig.rs
@@ -29,8 +29,8 @@ impl LspAdapter for ZlsAdapter {
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<Box<dyn 'static + Send + Any>> {
         let release =
-            latest_github_release("zigtools/zls", false, false, delegate.http_client()).await?;
-        let asset_name = format!("zls-{}-macos.tar.gz", ARCH);
+            latest_github_release("zigtools/zls", true, false, delegate.http_client()).await?;
+        let asset_name = format!("zls-{ARCH}-macos.tar.gz");
         let asset = release
             .assets
             .iter()

